source/s3_client.c

```diff
@@ -609,29 +609,54 @@ struct aws_s3_meta_request *aws_s3_client_make_meta_request(
     }
 
     struct aws_byte_cursor host_header_value;
+    AWS_ZERO_STRUCT(host_header_value);
 
-    if (aws_http_headers_get(message_headers, g_host_header_name, &host_header_value)) {
-        AWS_LOGF_ERROR(
-            AWS_LS_S3_CLIENT,
-            "id=%p Cannot create meta s3 request; message provided in options does not have a 'Host' header.",
-            (void *)client);
-        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
-        return NULL;
+    struct aws_string *endpoint_host_name = NULL;
+
+    if (aws_http_headers_get(message_headers, g_host_header_name, &host_header_value) == AWS_OP_SUCCESS) {
+        endpoint_host_name = aws_string_new_from_cursor(client->allocator, &host_header_value);
     }
 
     struct aws_s3_meta_request *meta_request = client->vtable->meta_request_factory(client, options);
 
     if (meta_request == NULL) {
         AWS_LOGF_ERROR(AWS_LS_S3_CLIENT, "id=%p: Could not create new meta request.", (void *)client);
+        if (endpoint_host_name) {
+            aws_string_destroy(endpoint_host_name);
+        }
+        return NULL;
+    }
+
+    bool is_https = true;
+    uint16_t port = 0;
+
+    if (options->uri != NULL) {
+        const struct aws_byte_cursor *uri_host_name_cursor = aws_uri_host_name(options->uri);
+        if (uri_host_name_cursor->len) {
+            struct aws_string *uri_host_name = aws_string_new_from_cursor(client->allocator, uri_host_name_cursor);
+            if (endpoint_host_name != NULL) {
+                aws_string_destroy(endpoint_host_name);
+            }
+            endpoint_host_name = uri_host_name;
+        }
+        struct aws_byte_cursor https_scheme = aws_byte_cursor_from_c_str("https");
+        is_https = aws_byte_cursor_eq(aws_uri_scheme(options->uri), &https_scheme);
+        port = aws_uri_port(options->uri);
+    }
+
+    if (endpoint_host_name == NULL) {
```

Comment:
Is it valid to have a request without host header for S3? Will server 500 us? I think we should still follow the logic before (Check the Host header exist or not).

Reply:
It is not valid - in this implementation - host header is required and is enforced to match uri if uri is supplied

Reply:
> host header is required and is enforced to match uri if uri is supplied

I don't think the code does any enforcement here, did I miss something?

Reply:
well - that is (currently) in the Java but as discussed above, manipulation of host header will be removed from the Java and the enforcement will be as discussed  [above](https://github.com/awslabs/aws-c-s3/pull/174#discussion_r807194886)

Reply:
changed in latest commit